connectors/stash/src/main/java/com/vmware/connectors/stash/StashController.java

```diff
@@ -0,0 +1,444 @@
+
+        final String url = String.format(StashConstants.STASH_PR_COMMENT_URL_FORMAT,
+                baseUrl,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId());
+
+        final StashComment stashComment = new StashComment(comment);
+
+        final ListenableFuture<ResponseEntity<String>> response = this.rest.exchange(
+                url,
+                HttpMethod.POST,
+                new HttpEntity<>(stashComment, headers),
+                String.class);
+        return Async.toSingle(response);
+    }
+
+    private List<String> getComments(final String baseUrl,
+                                     final HttpHeaders headers,
+                                     final StashPullRequest pullRequest) {
+        final String url = String.format(StashConstants.STASH_ACTIVITIES_URL_FORMAT,
+                baseUrl,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId());
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> response = this.rest.exchange(
+                url,
+                HttpMethod.GET,
+                new HttpEntity<>(headers),
+                JsonDocument.class);
+
+        final JsonDocument[] jsonDocuments = new JsonDocument[1];
+        Async.toSingle(response)
+                .subscribe(entity -> {
+                    jsonDocuments[0] = entity.getBody();
+                });
+        Assert.isTrue(Objects.nonNull(jsonDocuments[0]), "JsonDocument should not be null.");
+
+        final List<String> comments = jsonDocuments[0].read("$.values[*].comment.text");
+        return comments.stream()
+                .limit(COMMENTS_SIZE)
+                .collect(Collectors.toList());
+    }
+
+    private Single<ResponseEntity<String>> performStashAction(final String baseUrl,
+                                                              final String authHeader,
+                                                              final StashPullRequest pullRequest,
+                                                              final StashAction stashAction) {
+        final HttpHeaders headers = buildHeaders(authHeader);
+
+        // Get current version of the pull request. Pull request "version" changes when we do any actions on it.
+        // When the pull request is raised, the current value will be 0.
+        // For example, when we approve the pull request, then the version will change from 0 to 1.
+        // We have to add the latest version of the pull request URI to do any actions. Otherwise, the action will be rejected.
+        // If the build for the branch is going on, then the actions would be rejected.
+        final String version = getVersion(headers, baseUrl, pullRequest);
+
+        final String url = String.format(StashConstants.STASH_PR_ACTION_FORMAT,
```

Comment:
Again, UriComponentsBuilder is safer than String.format.

Reply:
I have used ```UriComponentsBuilder``` instead of ```String.format``` in all the places..